contracts/external/cw-fund-distributor/src/contract.rs

```diff
@@ -0,0 +1,628 @@
+use crate::error::ContractError;
+use crate::msg::{
+    CW20EntitlementResponse, CW20Response, DenomResponse, ExecuteMsg, InstantiateMsg, MigrateMsg,
+    NativeEntitlementResponse, QueryMsg, TotalPowerResponse, VotingContractResponse,
+};
+use crate::state::{
+    CW20_BALANCES, CW20_CLAIMS, DISTRIBUTION_HEIGHT, FUNDING_PERIOD_EXPIRATION, NATIVE_BALANCES,
+    NATIVE_CLAIMS, TOTAL_POWER, VOTING_CONTRACT,
+};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    to_binary, Addr, BankMsg, Binary, Coin, Decimal, Deps, DepsMut, Env, Fraction, MessageInfo,
+    Order, Response, StdError, StdResult, Uint128, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw_paginate::{paginate_map};
+use std::borrow::Borrow;
+use std::collections::HashMap;
+
+use dao_interface::voting;
+
+const CONTRACT_NAME: &str = "crates.io:cw-fund-distributor";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // store the height
+    DISTRIBUTION_HEIGHT.save(deps.storage, &env.block.height)?;
+
+    // get the funding expiration and store it
+    let funding_expiration_height = msg.funding_period.after(&env.block);
+    FUNDING_PERIOD_EXPIRATION.save(deps.storage, &funding_expiration_height)?;
+
+    // validate the contract and save it
+    let voting_contract = deps.api.addr_validate(&msg.voting_contract)?;
+    VOTING_CONTRACT.save(deps.storage, &voting_contract)?;
+
+    let total_power: voting::TotalPowerAtHeightResponse = deps.querier.query_wasm_smart(
+        voting_contract.clone(),
+        &voting::Query::TotalPowerAtHeight {
+            height: Some(env.block.height),
```

Comment:
We may want to make the distribution height a parameter for the `InstantiateMsg`.

imagine there is a proposal to dissolve a DAO with a large amount assets and a liquid token. Since everyone can see the proposal go live, they can rush to stake their tokens (or acquire some) before the proposal passes and is executed.

Likely not a huge issue, but also easy enough to add an optional `distribution_height` parameter.

Reply:
That's a great point. Initially I had it as a parameter but ended up removing it. Will get it back in there!